### PR TITLE
Fix for index error for Reduce Mean

### LINKED ISSUE #18898: Segmentation fault occurs  when loading ONNX v6 model shufflenet_v2
##### System information (version)

- OpenCV => 4.5.0
- Operating System / Platform => Ubuntu 20.04
- Compiler => gcc x64

##### Detailed description

when loading onnx net converted from Pytorch. segmentation fault occurs.
model:
https://pytorch.org/hub/pytorch_vision_shufflenet_v2/
onnx_file:

[net.zip](https://github.com/opencv/opencv/files/5582255/net.zip)

and map:

##### Steps to reproduce

```.cpp
cout << "net init begins!" << endl;
net = dnn::readNetFromONNX(modelBin);
cout << "Read done." << endl;
```
##### Issue submission checklist

 - [x] I report the issue, it's not a question
   
 - [x] I checked the problem with documentation, FAQ, open issues,
       answers.opencv.org, Stack Overflow, etc and have not found solution
   
 - [x] I updated to latest OpenCV version and the issue is still there
   
 - [x] There is reproducer code and related data files: videos, images, onnx, etc

### PR DESCRIPTION
**Pull request by joegeisbauer**
resolves #18898

### [ONNX] Fix Reduce Mean Layer Import Error

The onnx_importer.cpp was causing a Segmentation Fault without an error message when converting newer models like MobileNetV2 that had been converted from PyTorch torchvision using torch.onnx. This occurred with the recent update to OpenCV 4.5.0. The culprit was chased down to an indexing issue in the conversion of the ReduceMean layer. You can see the one line change I made in the code comparison of this commit. My understanding was that the intention was to delete the last two dimensions of the tensor which corresponded to the feature size in order to create the correct size tensor for output by changing the last two dimensions. If that is not the expected behavior, then this code will not fix the issue. However, we did verify that this code will fix the issue if that is the expected behavior.

<cut/>

### Pull Request Readiness Checklist

See details at https://github.com/opencv/opencv/wiki/How_to_contribute#making-a-good-pull-request

- [ X ] I agree to contribute to the project under Apache 2 License.
- [ X ] To the best of my knowledge, the proposed patch is not based on a code under GPL or other license that is incompatible with OpenCV
- [ X ] The PR is proposed to proper branch
- [ X ] There is reference to original bug report and related work
- [ X ] There is accuracy test, performance test and test data in opencv_extra repository, if applicable
      Patch to opencv_extra has the same branch name.
- [ X ] The feature is well documented and sample code can be built with the project CMake

```
force_builders=Custom,Custom Win,Custom Mac
build_image:Custom=ubuntu-openvino-2021.1.0:20.04
build_image:Custom Win=openvino-2021.1.0
build_image:Custom Mac=openvino-2021.1.0

test_modules:Custom=dnn,python2,python3,java
test_modules:Custom Win=dnn,python2,python3,java
test_modules:Custom Mac=dnn,python2,python3,java

buildworker:Custom=linux-1
# disabled due high memory usage: test_opencl:Custom=ON
test_opencl:Custom=OFF
test_bigdata:Custom=1
test_filter:Custom=*
```

### CODE REVIEW

**Comment by alalek** [comment]
@joegeisbauer Could you please share conversion instructions (with link on original model) and final .onnx file (thrugh GoogleDrive / OneDrive / DropBox), so we can reproduce the problem?

**Comment by joegeisbauer** [comment]
The models both the state dict checkpoint and the onyx files can be found [here](https://www.dropbox.com/sh/iuyfb95itbnh6xd/AADHLrd668ZR_tfo5d9xPa6ra?dl=0). For directions on model conversion, you can check out this [link](https://pytorch.org/tutorials/advanced/super_resolution_with_onnxruntime.html), but I have already converted it for you in the Dropbox folder. The following code was copied from the link

```
# Input to the model
x = torch.randn(batch_size, 1, 224, 224, requires_grad=True)
torch_out = torch_model(x)

# Export the model
torch.onnx.export(torch_model,               # model being run
                  x,                         # model input (or a tuple for multiple inputs)
                  "super_resolution.onnx",   # where to save the model (can be a file or file-like object)
                  export_params=True,        # store the trained parameter weights inside the model file
                  opset_version=10,          # the ONNX version to export the model to
                  do_constant_folding=True,  # whether to execute constant folding for optimization
                  input_names = ['input'],   # the model's input names
                  output_names = ['output'], # the model's output names
                  dynamic_axes={'input' : {0 : 'batch_size'},    # variable lenght axes
                                'output' : {0 : 'batch_size'}})
```

File: modules/dnn/src/onnx/onnx_importer.cpp [hunk 1]
```diff
@@ -496,7 +496,7 @@ void ONNXImporter::handleNode(const opencv_onnx::NodeProto& node_proto_)
                 bool keepdims = layerParams.get<int>("keepdims");
                 MatShape targetShape = inpShape;
                 for (int i = 0; i < axes.size(); i++) {
-                    int axis = clamp(axes.get<int>(i), inpShape.size());
+                    int axis = clamp(axes.get<int>(0), inpShape.size());
```
Comment (sl-sergei) [inline modules/dnn/src/onnx/onnx_importer.cpp:499]: This solution is slightly incorrect, it doesn't work with **keepdims**. I think it's better to try something like:
```
MatShape targetShape;
std::vector<bool> shouldDelete(inpShape.size(), false);
for (int i = 0; i < axes.size(); i++) {
    int axis = clamp(axes.get<int>(i), inpShape.size());
    shouldDelete[axis] = true;
}
for (int axis = 0; axis < inpShape.size(); ++axis){
    if (!shouldDelete[axis])
        targetShape.push_back(inpShape[axis]);
    else if (keepdims)
        targetShape.push_back(1);
}
```

Comment (sl-sergei) [reply modules/dnn/src/onnx/onnx_importer.cpp:499]: This way it doesn't brake the tests

**Comment by sl-sergei** [comment]
@joegeisbauer Can you please change the target branch from **master** to **3.4**?

File: modules/dnn/src/onnx/onnx_importer.cpp (hunk 1 above)
Comment (joegeisbauer) [reply modules/dnn/src/onnx/onnx_importer.cpp:499]: @sl-sergei Would you like me to make the changes, test, and then commit them to this pull request?

Comment (sl-sergei) [reply modules/dnn/src/onnx/onnx_importer.cpp:499]: > @sl-sergei Would you like me to make the changes, test, and then commit them to this pull request?

@joegeisbauer It was your initiative to make this contribution, and you did great work highlighting the problem, reproducing and proposing a fix. I think you should finish it:)

**Comment by asmorkalov** [comment]
This patch should [go](https://github.com/opencv/opencv/wiki/How_to_contribute#faq) into [3.4 branch](https://github.com/opencv/opencv/wiki/Branches) first. We will merge changes from 3.4 into master regularly (weekly/bi-weekly). 3.4 and master branches are diverged and it's not enough to change target in PR.

So, please:
- change "base" branch of this PR: master => 3.4 (use "Edit" button near PR title)
- rebase your commits from master onto 3.4 branch. For example:
    `git rebase -i --onto upstream/3.4 upstream/master`
    (check list of your commits, save and quit (Esc + "wq" + Enter)
    where `upstream` is configured by following [this GitHub guide](https://help.github.com/articles/configuring-a-remote-for-a-fork/) and fetched (`git fetch upstream`).
- push rebased commits into source branch of your fork (with `--force` option)

Note: no need to re-open PR, apply changes "inplace".

**Comment by joegeisbauer** [comment]
@asmorkalov I took care of the rebase as well. We will see how the builds and testing complete and go from there. Thank you @asmorkalov and @sl-sergei for your patience and understanding on my first ever pull request for OpenCV. It has been a great experience, and makes me want to help by contributing even more as I can.

**Review by sl-sergei — APPROVED** [review]
Thank you! :+1:

**Comment by alalek** [comment]
resolves #18898
